merkle_tree/src/main.rs

```diff
@@ -0,0 +1,432 @@
+        for proof in proofs {
+            if leaf_index % 2 == 1 {
+                // if index is odd we are on a left branch, so the verification must be computed concatenating the proof second
+                hash_for_verification = digest(hash_for_verification.to_owned() + &proof);
+            } else {
+                // if index is even we are on a right branch, so the verification must be computed concatenating the proof first
+                hash_for_verification = digest(proof.clone() + &hash_for_verification);
+                leaf_index -= 1; //if it's a right child this is necesary to calculate its parent
+            }
+            leaf_index = leaf_index / 2;
+        }
+        hash_for_verification == self.tree[0]
+    }
+
+    /// Given an index of a data hash the function must return the proof that the tree contains that data hash.
+    fn generate_proof(&self, mut node_index: usize) -> Option<Vec<String>> {
+        let mut proof: Vec<String> = vec![];
+        if node_index > self.tree.len() {
+            return None;
+        }
+        while node_index > 0 {
+            if node_index % 2 == 1 {
+                // if index is odd we are on a left branch, so the verification must be computed concatenating the proof second
+                proof.push(self.tree[node_index + 1].clone());
+            } else {
+                // if index is even we are on a right branch, so the verification must be computed concatenating the proof first
+                proof.push(self.tree[node_index - 1].clone());
+                node_index -= 1; //if it's a right child this is necesary to calculate its parent
+            }
+            node_index = node_index / 2;
+        }
+        Some(proof)
+    }
+
+    /// Returns a merkle tree with a new node, if the total amount of data is not a power of 2
+    /// the tree repeats the last leaf to build the base of the tree.
+    fn add_node(&mut self, data: &str) {
+        let mut new_data_vector: Vec<&str> = self.og_leaves.iter().map(String::as_str).collect();
+        new_data_vector.push(data);
+        let new_tree = Self::new(new_data_vector);
+        self.tree = new_tree.tree.clone();
+        self.leaves = new_tree.leaves;
+        self.og_leaves = new_tree.og_leaves.clone();
+    }
+}
+
+///Extends the size of a vector to a power of 2 by repeating the last value.
+fn extend_to_power2_size(vec: &Vec<String>) -> Vec<String> {
+    let mut copy: Vec<String> = vec.clone();
+    let diff_to_power_of_2 = ((vec.len() as f64).log2()).abs() - (vec.len()).ilog2() as f64;
+    if diff_to_power_of_2 != 0.0 {
+        let new_size = ((vec.len()).ilog2() as i32) * 4; // vec.len().ilog2 * 2 is the biggest power of 2 smaller than vec.len(), I multiply that by 2 to get the total amount of leaves
+        for _ in 0..(new_size as usize) - vec.len() {
+            copy.push(vec[vec.len() - 1].clone());
+        }
+    }
+    copy
+}
+
+fn main() {}
```

Comment:
Looks like the main function does nothing, how is the user expected to use this tree?